File: kcmjs/core/component.js
/**
 * Base component system for CUI framework
 */

export class Rect {
    constructor(x, y, width, height) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
}

export class Component {
    constructor(rect) {
        this.rect = rect;
        this.children = [];
        this.parent = null;
        this.visible = true;
        this.focused = false;
    }

    addChild(child) {
        child.parent = this;
        this.children.push(child);
    }

    removeChild(child) {
        const index = this.children.indexOf(child);
        if (index > -1) {
            child.parent = null;
            this.children.splice(index, 1);
        }
    }

    render(renderer) {
        // Default implementation: render children
        for (const child of this.children) {
            child.render(renderer);
        }
    }

    update(deltaTime) {
        for (const child of this.children) {
            child.update(deltaTime);
        }
    }

    handleKey(key) {
        // Try to handle key in children first
        for (const child of this.children) {
            if (child.handleKey && child.handleKey(key)) {
                return true;
            }
        }
        return false;
    }

    handleEvent(event) {
        for (const child of this.children) {
            if (child.handleEvent(event)) {
                return true;
            }
        }
        return false;
    }
}

File: kcmpy/core/component.py
"""Base component system for CUI framework."""
from typing import Optional, List
from dataclasses import dataclass


@dataclass
class Rect:
    """Rectangle bounds for component positioning."""
    x: int
    y: int
    width: int
    height: int


class Component:
    """Base class for all UI components."""
    
    def __init__(self, rect: Rect):
        self.rect = rect
        self.children: List[Component] = []
        self.parent: Optional[Component] = None
        self.visible = True
        self.focused = False
    
    def add_child(self, child: 'Component'):
        """Add child component."""
        child.parent = self
        self.children.append(child)
    
    def remove_child(self, child: 'Component'):
        """Remove child component."""
        if child in self.children:
            child.parent = None
            self.children.remove(child)
    
    def render(self, renderer):
        """Render component to screen. Override in subclasses."""
        # Default implementation: render all children
        for child in self.children:
            child.render(renderer)
    
    def update(self, delta_time: float):
        """Update component state."""
        for child in self.children:
            child.update(delta_time)
    
    def handle_event(self, event):
        """Handle input events."""
        for child in self.children:
            if child.handle_event(event):
                return True
        return False
